dotcom-rendering/src/useEffectReducer.ts

```diff
@@ -0,0 +1,79 @@
+import type { Dispatch } from 'react';
+import { useEffect, useReducer, useRef } from 'react';
+
+type StateAndEffect<State, Effect> = {
+	state: State;
+	effect?: Effect;
+};
+
+/**
+ * @param reducer Very similar to the reducer normally passed to
+ * {@linkcode React.useReducer}, but also returns a description of effects
+ * alongside the state.
+ *
+ * @param effectReducer A handler for effects. It takes a description of an
+ * effect and decides how to run that effect, returning an `Action` that will
+ * be fed back into the next loop of the `reducer`.
+ *
+ * Depending on how an application is written, different `effectReducer`s can
+ * be passed in different environments. So in production this might run actual
+ * effects, whereas in tests it might run mocked versions.
+ *
+ * *Note:* The value used will be the one passed on the initial render; the
+ * argument is ignored after this point.
+ *
+ * @param initialState The state at the start of the application; the same
+ * as the `initialState` argument usually passed to
+ * {@linkcode React.useReducer}.
+ *
+ * @param initialEffect Optionally, an effect to run at the start of the
+ * application.
+ *
+ * @returns State and a `dispatch` function, the same as
+ * {@linkcode React.useReducer}.
+ */
+const useEffectReducer = <State, Effect, Action>(
+	reducer: (
+		state: State,
+		action: Action,
+	) => { state: State; effect?: Effect },
+	effectReducer: (effect: Effect) => Action | Promise<Action>,
```

Comment:
What do people think about using the term "reducer" for this? On the one hand it has symmetry with the `reducer` parameter, but on the other it doesn't really match the structure of a function that would be passed to `reduce` (which I believe is the reason the term "reducer" is used):

```ts
actions.reduce((state, action) => state, initialState)
```

It's closer to a function you'd pass to `map`, albeit side-effectful:

```ts
effects.map(effect => action)
```